L1Trigger/TrackTrigger/src/TrackQuality.cc

```diff
@@ -0,0 +1,298 @@
+  std::vector<edm::Ref<edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>, TTStub<Ref_Phase2TrackerDigi_>>> stubRefs =
+      aTrack.getStubRefs();
+  float tmp_trk_nstub = stubRefs.size();
+
+  // While not strictly necessary to define these parameters,
+  // it is included so each variable is named to avoid confusion
+  float tmp_trk_big_invr = 500 * abs(aTrack.rInv());
+  float tmp_trk_tanl = abs(aTrack.tanL());
+  float tmp_trk_z0 = aTrack.z0();
+  float tmp_trk_phi = aTrack.phi();
+  float tmp_trk_pt = aTrack.momentum().perp();
+  float tmp_trk_eta = aTrack.eta();
+  float tmp_trk_chi2 = aTrack.chi2();
+  float tmp_trk_log_chi2 = log(tmp_trk_chi2);
+  float tmp_trk_log_chi2rphi = log(tmp_trk_chi2rphi);
+  float tmp_trk_log_chi2rz = log(tmp_trk_chi2rz);
+  float tmp_trk_log_bendchi2 = log(tmp_trk_bendchi2);
+
+  // fill feature map
+  std::map<std::string, float> feature_map;
+  feature_map["log_chi2"] = tmp_trk_log_chi2;
+  feature_map["log_chi2rphi"] = tmp_trk_log_chi2rphi;
+  feature_map["log_chi2rz"] = tmp_trk_log_chi2rz;
+  feature_map["log_bendchi2"] = tmp_trk_log_bendchi2;
+  feature_map["chi2"] = tmp_trk_chi2;
+  feature_map["chi2rphi"] = tmp_trk_chi2rphi;
+  feature_map["chi2rz"] = tmp_trk_chi2rz;
+  feature_map["bendchi2"] = tmp_trk_bendchi2;
+  feature_map["nstub"] = tmp_trk_nstub;
+  feature_map["lay1_hits"] = float(hitpattern_expanded_binary[0]);
+  feature_map["lay2_hits"] = float(hitpattern_expanded_binary[1]);
+  feature_map["lay3_hits"] = float(hitpattern_expanded_binary[2]);
+  feature_map["lay4_hits"] = float(hitpattern_expanded_binary[3]);
+  feature_map["lay5_hits"] = float(hitpattern_expanded_binary[4]);
+  feature_map["lay6_hits"] = float(hitpattern_expanded_binary[5]);
+  feature_map["disk1_hits"] = float(hitpattern_expanded_binary[6]);
+  feature_map["disk2_hits"] = float(hitpattern_expanded_binary[7]);
+  feature_map["disk3_hits"] = float(hitpattern_expanded_binary[8]);
+  feature_map["disk4_hits"] = float(hitpattern_expanded_binary[9]);
+  feature_map["disk5_hits"] = float(hitpattern_expanded_binary[10]);
+  feature_map["rinv"] = tmp_trk_big_invr;
+  feature_map["tanl"] = tmp_trk_tanl;
+  feature_map["z0"] = tmp_trk_z0;
+  feature_map["phi"] = tmp_trk_phi;
+  feature_map["dtot"] = float(tmp_trk_dtot);
+  feature_map["ltot"] = float(tmp_trk_ltot);
+  feature_map["pt"] = tmp_trk_pt;
+  feature_map["eta"] = tmp_trk_eta;
+  feature_map["nlaymiss_interior"] = float(tmp_trk_nlaymiss_interior);
+  feature_map["bendchi2_bin"] = tmp_trk_bendchi2_bin;
+  feature_map["chi2rphi_bin"] = tmp_trk_chi2rphi_bin;
+  feature_map["chi2rz_bin"] = tmp_trk_chi2rz_bin;
+
+  // fill tensor with track params
+  transformedFeatures.reserve(featureNames.size());
+  for (const std::string& feature : featureNames)
+    transformedFeatures.push_back(feature_map[feature]);
+
+  return transformedFeatures;
+}
+
+void TrackQuality::setTrackQuality(TTTrack<Ref_Phase2TrackerDigi_>& aTrack) {
+  if (this->qualityAlgorithm_ == "Cut") {
```

Comment:
doing multiple string comparisons for every track is suboptimal
consider using enum class (or a factory, though that might take more work)

Reply:
done using enum class, a string comparison is used to set the enum class of the track quality but this is performed once per event rather than once per track